dlab_core/infrastructure/services.py

```diff
@@ -0,0 +1,187 @@
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+# ******************************************************************************
+import requests
+import urllib3
+
+from dlab_core.domain.exceptions import DLabException
+from dlab_core.domain.helper import validate_property_type
+
+from jose import jwt
+
+
+URL_INTROSPECT = '{realm_address}/protocol/openid-connect/token/introspect'
+
+
+urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
+
+
+class KeyCloakException(DLabException):
+    pass
+
+
+class KeyCloakConnectionException(KeyCloakException):
+    pass
+
+
+class KeyCloak(object):
+    """Client for KeyCloak interaction
+
+    :type keycloak_host: str
+    :param keycloak_host: IP or hostname for KeyCloak server to use
+
+    :type realm_name: str
+    :param realm_name: name of the KeyCloak Realm to use
+
+    :type client_id: str
+    :param client_id: KeyCloak Client to use for token introspection
+
+    :type client_secret: str
+    :param client_secret: KeyCloak Client secret string
+    """
+    def __init__(self, keycloak_host, realm_name, client_id, client_secret):
+        if not (isinstance(keycloak_host, str) and isinstance(realm_name, str)):
+            raise DLabException(
+                'Either Keycloak host or Realm name '
+                'has invalid type ({}, {})'.format(keycloak_host, realm_name)
+            )
+        self.realm_address = (keycloak_host, realm_name)
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self._pub_key = None
+
+    @property
+    def realm_address(self):
+        return self._realm_address
+
+    @realm_address.setter
+    def realm_address(self, value):
+        self._realm_address = '{}/realms/{}'.format(
+            value[0].rstrip('/'), value[1]
+        )
+
+    @property
+    def client_id(self):
+        return self._client_id
+
+    @client_id.setter
+    @validate_property_type(str)
+    def client_id(self, value):
+        self._client_id = value
+
+    @property
+    def client_secret(self):
+        return self._client_secret
+
+    @client_secret.setter
+    @validate_property_type(str)
+    def client_secret(self, value):
+        self._client_secret = value
+
+    @property
+    def public_key(self):
+        if not self._pub_key:
+            self._pub_key = self.retrieve_key()
+
+        return self._pub_key
+
+    def make_request_to_server(self, method='GET', endpoint='', verify=False,
```

Comment:
why do you need default value for endpoint, is it optional parameter and you are going to perform request to empty string?

Reply:
Removed default value